### ISSUE TITLE
Add more columns to summary output?

### ISSUE THREAD

**Issue by koheiw** [quanteda/quanteda#1912]
I have a function that counts the number of non-word element. I used this to identify documents with a lot of noise (and drop them). How about returning these columns from `summary.corpus`? I expect that additional cost is almost nothing as tokenization is the main cost. 
```r
diagnosys.corpus <- function(x, ...) {

    dict <- dictionary(list(
        "number" = "\\p{N}",
        "punct" = "\\p{P}",
        "symbol" = "\\p{S}",
        "any" = "[\\p{N}\\p{P}\\p{S}]"
    ))

    n_sent <- ntoken(tokens(x, what = "sentence"))
    temp <- dfm(tokens(x, ...))
    n_token <- ntoken(temp)
    result <- convert(
        dfm_lookup(temp, dictionary = dict, valuetype = "regex"),
        "data.frame"
    )
    result$n_sent <- n_sent
    result$n_token <- n_token
    result$dupli <- duplicated(texts(x))
    result$noise <- result$any / result$n_token
    return(result)
}
```

**Comment by kbenoit** [comment]
This is such a good idea that... it's already available!

``` r
library("quanteda")
## Package version: 2.0.0.9000

corp <- corpus(data_char_ukimmig2010) %>%
  quanteda:::add_summary_metadata(extended = TRUE)

meta(corp, type = "system")
## $`package-version`
## [1] '2.0.0.9000'
## 
## $`r-version`
## [1] '3.6.3'
## 
## $system
##   sysname   machine      user 
##  "Darwin"  "x86_64" "kbenoit" 
## 
## $directory
## [1] "/private/var/folders/1v/ps2x_tvd0yg0lypdlshg_vwc0000gp/T/Rtmphgbqcv/reprexd50879c8e658"
## 
## $created
## [1] "2020-03-12"
## 
## $summary
## Corpus consisting of 9 documents, showing 9 documents:
## 
##          Text Types Tokens Sentences
##           BNP  1125   3280        88
##     Coalition   142    260         4
##  Conservative   251    499        15
##        Greens   322    677        21
##        Labour   298    680        29
##        LibDem   251    483        14
##            PC    77    114         5
##           SNP    88    134         4
##          UKIP   346    722        27
## 
## 
## $summary_extended
## $summary_extended$total_tokens
## [1] 6849
## 
## $summary_extended$all_tokens
##          BNP    Coalition Conservative       Greens       Labour       LibDem 
##         3280          260          499          677          680          483 
##           PC          SNP         UKIP 
##          114          134          722 
## 
## $summary_extended$total_punctuation
## [1] 792
## 
## $summary_extended$total_symbols
## [1] 0
## 
## $summary_extended$total_numbers
## [1] 63
## 
## $summary_extended$total_words
## [1] 5994
## 
## $summary_extended$total_stopwords
## [1] 97
## 
## $summary_extended$top_dfm
## Document-feature matrix of: 1 document, 100 features (0.0% sparse).
##     features
## docs immigration bnp can british people become well within years identity
##    1          66  13  11      37     35      6    9     11    11        6
## [ reached max_nfeat ... 90 more features ]
```

But we would need to add this output to summary(). Maybe we should give users the option of compiling this information when creating a new corpus.

**Comment by koheiw** [comment]
Very good. I think word and non-word tokens should be counted separately. There are people who are using the number in Tokens column for analysis...

**Comment by kbenoit** [comment]
Well we do have the ability to pass arguments to `ntokens()` and `ntype()` via ... in `summary(x, ...)` but no one apparently uses that. So I agree, separating them is a good idea.

**Comment by koheiw** [comment]
I started thinking how and whether we can implement caching mechanism for summary. It is simple, but I am not sure if we are allowed to modify objects in parent frames... If it is allowed, caching would speed us many other functions.
```r
# Make corpus as usual
> corp <- corpus(c("aaaa aaaa.", "bbbb! 123"))

# There is not cached summary yet
> attr(corp, "meta")[["system"]]["summary"]
$<NA>
NULL

# Summarize and cache
> summary(corp)
  doc_id number punct symbol any n_sent n_token n_type dupli     noise
1  text1      0     1      0   1      1       3      5 FALSE 0.3333333
2  text2      1     1      0   2      2       3      5 FALSE 0.6666667

# Confirm that it is cached
> attr(corp, "meta")[["system"]]["summary"]
$summary
$summary$hash
[1] "dfe4a284584fa05b69b72e00e532f72c1906198a5a8c108d05d919a3a4efd6b0"

$summary$data
  doc_id number punct symbol any n_sent n_token n_type dupli     noise
1  text1      0     1      0   1      1       3      5 FALSE 0.3333333
2  text2      1     1      0   2      2       3      5 FALSE 0.6666667
```

**Comment by kbenoit** [comment]
why not just do it by default when `corpus()` is called? Modifying parent frames is the dark arts and might have unexpected (negative) consequences for programmers using our functions inside their code, functions, or packages. But it's worth exploring.

**Comment by koheiw** [comment]
I checked the code of **fastmatch**. It does it in Rcpp as R objects are passed by reference. We could take a similar approach...

**Comment by koheiw** [comment]
It seems possible without the dark arts.

```r
> corp <- corpus(c("aaaa aaaa.", "bbbb! 123"))
> 
> # There is not cached summary yet
> attr(corp, "meta")[["object"]]["summary"]
$<NA>
NULL

> 
> # Summarize and cache
> summary(corp)
  doc_id number punct symbol any n_sent n_token n_type dupli     noise
1  text1      0     1      0   1      1       3      5 FALSE 0.3333333
2  text2      1     1      0   2      2       3      5 FALSE 0.6666667
> 
> # Confirm that it is cached
> attr(corp, "meta")[["object"]]["summary"]
$summary
$summary$hash
[1] "d87bc84d14fcc2e1d0755ea8d30d0ec5bf09218b3b284a0797ba86d0e33d7c4d"

$summary$data
  doc_id number punct symbol any n_sent n_token n_type dupli     noise
1  text1      0     1      0   1      1       3      5 FALSE 0.3333333
2  text2      1     1      0   2      2       3      5 FALSE 0.6666667


> 
> # Don't cahce
> summary(corp, cache = FALSE)
  doc_id number punct symbol any n_sent n_token n_type dupli     noise
1  text1      0     1      0   1      1       3      5 FALSE 0.3333333
2  text2      1     1      0   2      2       3      5 FALSE 0.6666667
> attr(corp, "meta")[["object"]]["summary"]
NULL
```
We only need 
```c
// [[Rcpp::export]]
void qatd_cpp_set_meta(CharacterVector texts_, List meta_) {
    texts_.attr("meta") = meta_;
}
```

**Comment by kbenoit** [comment]
Cool. What if we pass that through to tokens, and subsequent `tokens()` calls that change the tokens would update the summary info? Like removing punctuation. Then we could add `summary.tokens()`.

**Comment by koheiw** [comment]
Adding `summary.tokens()` and even `summary.dfm()` is a natural things for us to do. Object meta is passed to downstream object, but need to be reset once the main object is modified (detected by the hash value).

This part is same for all the objects.
https://github.com/quanteda/quanteda/blob/16ee121fd4c618f6d0fb33f4bc33922acfb6eb63/R/character-methods.R#L22-L28

**Comment by koheiw** [comment]
I will try issue a PR soon but not sure where `summary.coprus`, `summary.tokens`, `summary.dfm` be documented? Generic summary page is already messy with with base objects... I started feeling that the function should be named differently as it is becoming more quanteda specific.

**Comment by kbenoit** [comment]
I'd suggest we document it as `summary-quanteda`, similar to how we do for `print-quanteda`. This shows up one place in the index, but `?summary.corpus` and `?summary.tokens` etc will raise that man page. And we can link to it from other documentation, such as the `corpus` help.

**Comment by koheiw** [comment]
Closed via #1955
